### ISSUE TITLE
Improved support for covenants

### ISSUE THREAD

**Issue by rkalis** [Bitcoin-com/cashscript#21]
Covenants are one of the main ways to add more useful functionality to Bitcoin Cash contracts (see https://honest.cash/v2/pein_sama/spending-constraints-with-op_checkdatasig-172). The way they work is by passing in the current transaction's preimage and using OP_CHECKDATASIG to ensure that it is the correct preimage. The components of this preimage can then be used in the rest of the script. For example, [Karol's Mecenas contract](https://github.com/KarolTrzeszczkowski/Mecenas-recurring-payment-EC-plugin) checks that the outputs and value of the transaction matches expectations.

It is a lot of work to manually recreate this transaction preimage and pass it in as a parameter. A first option could be to provide a placeholder for this preimage in the same way we offer a placeholder for `new Sig(keypair, 0x01`. We could take it further by adding language constructs for the different parts of the preimage, such as `tx.value`, or `tx.outputs`, and then passing in the preimage behind the scenes in the SDK. This needs more research though, and might need other language features to be completed first, such as simple arrays.

Another thing is that covenants can rely on a certain inputs/outputs setup in order to function, that possibly can't be guaranteed by the current CashScript SDK, since this does input / output creation behind the scenes. So either we will need to come up with a higher level of abstraction to make sure these inputs / outputs are set up correctly behind the scenes, or we need to offer a lower level of abstraction where the user creates inputs and outputs (similar to what the Spedn SDK does currently).

**Comment by rkalis** [comment]
A first few steps for higher abstraction on the side of the compiler could be:

- [ ] Add optional automatic bytes size enforcement. If a specific flag is set, we could add `<depth> OP_PICK OP_SIZE <size> OP_EQUALVERIFY` to the compiled output to make sure that the passed arguments are the correct size. Or `<depth> OP_PICK 4 OP_NUM2BIN OP_DROP` to make sure an `int` is actually numeric. Currently, this has to be done manually to achieve this kind of security.

- [ ] Add automatic output construction. Currently outputs are constructed manually, like this:
  ```solidity
  bytes output = amount + 0x17a914 + hash160(contractBytecode) + 0x87;
  ```
  It would be more efficient and legible to create this kind of functionality:
  ```solidity
  bytes output = OutputP2SH(amount, hash160(contractBytecode));
  ```

**Comment by rkalis** [comment]
Released better covenant support in `v0.3.0`. Something that could be added in the future is manual input selection to make sure that contracts that depend on it can work correctly.
